src/extended_json.ts

```diff
@@ -75,8 +75,9 @@ function deserializeValue(value: any, options: EJSON.Options = {}) {
 
     // if it's an integer, should interpret as smallest BSON integer
     // that can represent it exactly. (if out of range, interpret as double.)
-    if (Math.floor(value) === value) {
+    if (Number.isInteger(value) && !Object.is(value, -0)) {
       if (value >= BSON_INT32_MIN && value <= BSON_INT32_MAX) return new Int32(value);
+      // TODO(NODE-4377): EJSON js number handling diverges from BSON
```

Comment:
Do we have a test for this behavior somewhere?  If not, should we add a test demonstrating now that `-0` will deserialize to a double instead of an integer?

Reply:
added!